### PR TITLE
feat: setup authentication

### PR DESCRIPTION
**Pull request by munozlucas**
[Trello ticket](https://trello.com/c/07grdwlY/32-authentication-back)

- setup authentication module with Passport JWT
- create admin user on application bootstrap
- create signIn endpoint

### CODE REVIEW

File: src/app.service.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    @Inject(authConfig.KEY)
+    private readonly authConf: ConfigType<typeof authConfig>,
+    private readonly adminRepository: AdminRepository,
+  ) {}
+
+  async onApplicationBootstrap() {
+    const admin = await this.adminRepository.findByUsername(
+      this.authConf.adminUsername,
+    );
+    if (!admin && true) {
```
Comment (marianozunino) [inline src/app.service.ts:18]:
```suggestion
    if (!admin && true) {
```
Why do we need that `true` value?

File: src/auth/auth.module.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+@Module({
+  imports: [
+    RepositoriesModule,
+    JwtModule.registerAsync({
+      useFactory: async (
+        authConf: ConfigType<typeof authConfig>,
+      ): Promise<object> => ({
+        secret: authConf.jwtSecret,
+        signOptions: {
+          expiresIn: '1d',
```
Comment (marianozunino) [inline src/auth/auth.module.ts:19]:
```suggestion
          expiresIn: '1d',
```
We should have this as part of the `authConfig`

File: src/auth/strategies/jwt.strategy.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    super({
+      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+      ignoreExpiration: false,
+      secretOrKey: authConf.jwtSecret,
+    });
+  }
+
+  async validate(payload: any) {
+    return { userId: payload.sub, username: payload.username };
+  }
```
Comment (marianozunino) [inline src/auth/strategies/jwt.strategy.ts:22]: Not needed right now, but I would be cool if we have types for the `payload` param and the return type as well.

File: src/repositories/admin.repository.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import { AdminEntity } from '..//entities/admin.entity';
+import { Repository } from 'typeorm';
+import bcrypt from 'bcrypt';
+
+@Injectable()
+export class AdminRepository {
+  constructor(
+    @InjectRepository(AdminEntity)
+    private adminRepository: Repository<AdminEntity>,
+  ) {}
```
Comment (marianozunino) [inline src/repositories/admin.repository.ts:13]: This is ok, but using `TypeORM` allows you to inject the repository pattern without creating your own repo layer.

We can ditch this layer or go with it. Up to the team.

File: src/app.service.ts (hunk 1 above)
Comment (munozlucas) [reply src/app.service.ts:18]: this is something that shouldn't be there :see_no_evil:

File: src/auth/strategies/jwt.strategy.ts (hunk 3 above)
Comment (munozlucas) [reply src/auth/strategies/jwt.strategy.ts:22]: Yes, I have a separated task to type everything :+1:

File: package.json [hunk 5]
```diff
@@ -66,23 +75,6 @@
...
-    "testRegex": ".*\\.spec\\.ts$",
-    "transform": {
-      "^.+\\.(t|j)s$": "ts-jest"
-    },
-    "collectCoverageFrom": [
-      "**/*.(t|j)s"
-    ],
-    "coverageDirectory": "../coverage",
-    "testEnvironment": "node"
-  },
```
Comment (marianozunino) [inline package.json:85]: <3

**Review by marianozunino — APPROVED** [review]

**Pull request merged**